### DISCUSSION TITLE
Better support for testing Listeners with `Queue::fake()`

### DISCUSSION THREAD

**Discussion by calebdw** [laravel/framework#47871] [Ideas]
### Laravel Version

9.52.x

### PHP Version

8.1.x

### Database Driver & Version

Postgres 14

### Description

Hello!

It is very non-trivial to test Listeners with QueueFake at the moment; mostly because the Listeners are inside the `CallQueuedListener` job.

Therefore, asserting whether or not a Listener was pushed requires something like:
```php
function assertListenerPushed(string $listener): void
{
    Queue::assertPushed(fn (CallQueuedListener $job) => $job->class === $listener);
}

function assertListenerNotPushed(string $listener): void
{
    Queue::assertNotPushed(fn (CallQueuedListener $job) => $job->class === $listener);
}
```
which either requires duplicated code or a custom function (which is not ideal).

The issue I'm running into at the moment is that I have several Listeners listening to a specific Event, and I want to fake all but one of them (the one I'm currently testing). However, doing the following does not work:

```php
    Queue::fake()->except([
        <listener>
    ]);
```
as everything is faked because `<listener>` is not an actual job---I'm not sure if it's possible to actually fake a listener without overriding `QueueFake::shouldDispatchJob()`.

All in all, it would be very helpful to include first class testing support for Listeners just like with Jobs.

Thanks!

### Steps To Reproduce

n/a

**Comment by i-bajrai** [comment]
@calebdw I'm finding that I'm agreeing with your perspective here, using `CallQueuedListener` I feel is a workaround or a hack.

I used the following code to remove the event listeners I wasn't interested in testing, can anyone explain why this might be a bad idea or perhaps the approach is ok?

`MY_JOB_TO_BE_QUEUED` is inside of the `CheckCreditSpentListener` and will only be called if `shouldQueue` passes.

```
protected function isolateEventListener(): void
{
        Event::forget(TransactionWasCreated::class);
        Event::listen(TransactionWasCreated::class, CheckCreditSpentListener::class);
}
```

```
Queue::fake([
    MY_JOB_TO_BE_QUEUED::class,
]);

$this->isolateEventListener();

// Insert code to simulate the environment that the job will NOT be queued

$event = new TransactionWasCreated();

event($event);

Queue::assertNotPushed(MY_JOB_TO_BE_QUEUED::class);
```

**Comment by jeanlucnguyen** [comment]
This is definitely a bug, as if we read the documentation https://laravel.com/docs/9.x/mocking#queue-fake, we should be able to use our listener directly without having to go through the hack with `CallQueuedListener`.

Also, the current code prevents completely the use of the `except` method, or the `$jobsToFake` argument in the constructor of `QueueFake`, as they will never be matched, all the listeners being wrapped in `CallQueuedListener`

I've made the fix here:
https://github.com/laravel/framework/pull/54056/files
